gh-134559: Expand and reorganise documentation for the ``copy`` module (#134695)
python/cpython: Doc/library/copy.rst

between a target and an object. For collections that are mutable or contain
mutable items, a copy is sometimes needed so one can change one copy without
changing the other. This module provides generic shallow and deep copy
operations (explained below), plus a replace (copy-with-modifications)
operation that's implemented for certain immutable types.

> ``copy`` and ``deepcopy``
> -------------------------

Reviewer: ```suggestion
:func:`!copy()` and :func:`!deepcopy()`
---------------------------------------
```

   Return a deep copy of *obj*.

.. _shallow_vs_deep_copy:

.. tip:: Shallow vs. Deep Copy

>    The difference between shallow and deep copying is only relevant for compound
>    objects (objects that contain other objects, like lists or class instances):

Reviewer 2: 'container' is more commonly used and the name of a collections ABC
```suggestion
   The difference between shallow and deep copying is only relevant for container
   objects; those that contain other objects, like lists or class instances:
```

   Called to implement the deep copy operation; it is passed one
   argument, the *memo* dictionary.  If the ``__deepcopy__`` implementation needs
   to make a deep copy of a component, it should call the :func:`!copy.deepcopy` function
   with the component as first argument and the *memo* dictionary as second argument.
   The *memo* dictionary should be treated as an opaque object.

> ``replace``
> -----------

Reviewer 2: ```suggestion
:func:`!replace()`
------------------
```

.. function:: copy.replace(obj, /, **changes)

   Return a new object of the same type as *obj*, replacing fields with values
   from *changes*.

>    Function :func:`!copy.replace` is more limited
>    than :func:`~copy.copy` and :func:`~copy.deepcopy`,
>    and only supports named tuples created by :func:`~collections.namedtuple`,
>    :mod:`dataclasses`, and other classes which implement the replace protocol
>    (i.e. define a method :meth:`~object.__replace__`).

Reviewer 2: ```suggestion
   The :func:`!replace` function is more limited than 
   :func:`~copy.copy` and :func:`~copy.deepcopy`.
   For example, it only supports named tuples created by :func:`~collections.namedtuple`,
   :mod:`dataclasses`, and other classes which implement the replace protocol
   (i.e. those defining a :meth:`~object.__replace__` method).
```

>    (i.e. define a method :meth:`~object.__replace__`).

Reviewer: ```suggestion
   (that is, those defining a :meth:`~object.__replace__` method).
```
